hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfersMarshal.java

```diff
@@ -66,17 +74,117 @@ public ImpliedTransfers unmarshalFromGrpc(CryptoTransferTransactionBody op) {
 		}
 
 		final List<BalanceChange> changes = new ArrayList<>();
+		final List<Pair<EntityId, List<CustomFee>>> entityCustomFees = new ArrayList<>();
+		List<CustomFeesBalanceChange> customFeeBalanceChangesForRecord = new ArrayList<>();
 		for (var aa : op.getTransfers().getAccountAmountsList()) {
 			changes.add(hbarAdjust(aa));
 		}
+		var payerId = EntityId.fromGrpcAccountId(payer);
 		for (var scopedTransfers : op.getTokenTransfersList()) {
 			final var grpcTokenId = scopedTransfers.getToken();
-			final var scopingToken = Id.fromGrpcToken(grpcTokenId);
+			final var scopingToken = EntityId.fromGrpcTokenId(grpcTokenId);
+			var amount = 0L;
 			for (var aa : scopedTransfers.getTransfersList()) {
 				changes.add(tokenAdjust(scopingToken, grpcTokenId, aa));
+				if (aa.getAmount() > 0) {
+					amount += aa.getAmount();
+				}
+			}
+
+			List<CustomFee> customFeesOfToken = customFeeSchedules.lookupScheduleFor(scopingToken);
+			entityCustomFees.add(new Pair<>(scopingToken, customFeesOfToken));
+			List<BalanceChange> customFeeChanges = computeBalanceChangeForCustomFee(scopingToken, payerId, amount,
+					customFeesOfToken);
+			changes.addAll(customFeeChanges);
+			customFeeBalanceChangesForRecord.addAll(getListOfBalanceChangesForCustomFees(customFeeChanges));
+		}
```

Comment:
One question regarding the custom fee change calculation. Here is the scenario: A transfer will transfer token A and B to other accounts. A has its custom fee schedules which require to transfer token C to some account. C again has its custom fee schedule.

The question is, shall we calculate the balance changes recursively?

Reply:
This is a good question @ljianghedera. I think we should do only simple transfers and charges. A mutually recursive transfer (transfer of token A requires custom fee of B and transfer of token B also requires custom fee of A), if not being handled as mentioned, will complicate a lot in the fee calculations and also in our understandings/interpretations.

Reply:
Yeah, it is. So I think we need to stick to current implementation, and we may need to document this to address potential confusions and questions.